src/uct/ib/mlx5/ib_mlx5_hw.c

```diff
@@ -0,0 +1,218 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2018.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+#include <config.h>
+#if HAVE_INFINIBAND_MLX5_HW_H
+#include <infiniband/mlx5_hw.h>
+#include <ucs/arch/bitops.h>
+#include "ib_mlx5_log.h"
```

Comment:
need to put "" includes before <> includes (code style)

Reply:
#include "ib_mlx5_dv.h" is a special case. It contains declaration common for ib_mlx5_dv.c and ib_mlx5_hw.c and relays on mlx5 definitions (like mlx5_wqe_av, mlx5_cqe64). So, depending on which .c file it was included from it should take those definitions either from infiniband/mlx5dv.h or infiniband/mlx5_hw.h. I want to put the include before it, but this violates code style. Can we make exception in this case?

Reply:
can "ib_mlx5_dv.h" include either infiniband/mlx5dv.h or infiniband/mlx5_hw.h itself, according to config.h settings?
if not possible, need to prevent "ib_mlx5_dv.h" from being included from anywhere else except {ib_mlx5_dv.c and ib_mlx5_hw.c} (or at least comment it clearly)
BTW, can we move "ib_mlx5_hw.h" and "ib_mlx5_log.h" includes to the beginning of the file?

Reply:
The problem is: 
infiniband/mlx5dv.h and infiniband/mlx5_hw.h cannot be included in one .c file
ib_mlx5_hw.c is the only one which includes mlx5_hw.h in presence of mlx5dv.h in the system.
Hence none of files included from ib_mlx5_hw.c (they are ib_mlx5_dv.h - common interface and ib_mlx5_log.h) cannot include (directly or indirectly) mlx5dv.h though both depended on definitions from infiniband/mlx5* files.
ib_mlx5_dv.h is included inside ib_mlx5.h immediately after one of infiniband/mlx5*.
https://github.com/openucx/ucx/pull/2695/commits/064f398078f0c7a0a9118df51141b0772e2ea54f#diff-5441263d415c4a16ea2406a91aa25c85R40

Reply:
"ib_mlx5_log.h" should be removed

Reply:
"ib_mlx5_hw.h" can be moved up

Reply:
need to add include guard to "ib_mlx5_dv.h", so it could only be included from either ib_mlx5.h or ib_mlx5_hw.c and document why